x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_prebuilt_rules_upgrade_state.ts

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useMemo, useState } from 'react';
+import type { FieldsDiff } from '../../../../../../common/api/detection_engine';
+import {
+  ThreeWayDiffConflict,
+  type DiffableAllFields,
+  type DiffableRule,
+  type RuleObjectId,
+  type RuleSignatureId,
+  type RuleUpgradeInfoForReview,
+} from '../../../../../../common/api/detection_engine';
+import { convertRuleToDiffable } from '../../../../../../common/detection_engine/prebuilt_rules/diff/convert_rule_to_diffable';
+
+export interface RuleUpgradeState extends RuleUpgradeInfoForReview {
+  /**
+   * Rule containing desired values users expect to see in the upgraded rule.
+   */
+  finalRule: DiffableRule;
+  /**
+   * Indicates whether there are conflicts blocking rule upgrading.
+   */
+  hasUnresolvedConflicts: boolean;
+}
+export type RulesUpgradeState = Record<RuleSignatureId, RuleUpgradeState>;
+export type SetFieldResolvedValueFn<
+  FieldName extends keyof DiffableAllFields = keyof DiffableAllFields
+> = (params: {
+  ruleId: RuleObjectId;
+  fieldName: FieldName;
+  resolvedValue: DiffableAllFields[FieldName];
+}) => void;
+
+type RuleResolvedConflicts = Partial<DiffableAllFields>;
+type RulesResolvedConflicts = Record<string, RuleResolvedConflicts>;
+
+interface UseRulesUpgradeStateResult {
+  rulesUpgradeState: RulesUpgradeState;
+  setFieldResolvedValue: SetFieldResolvedValueFn;
+}
+
+export function usePrebuiltRulesUpgradeState(
+  ruleUpgradeInfos: RuleUpgradeInfoForReview[]
+): UseRulesUpgradeStateResult {
+  const [rulesResolvedConflicts, setRulesResolvedConflicts] = useState<RulesResolvedConflicts>({});
```

Comment:
I wonder if we should also let the user revert a conflict back to unresolved. Let's say they set some resolved value and then decide that it's not good and they want to start over from the initial state (merge proposal). In current implementation even if you close and reopen the flyout the state will be preserved. And if we later add persistence in `sessionStorage` even page reload won't help. 

Should we then think about adding buttons for reverting? @maximpn @jpdjere

Reply:
Yes. It might be helpful feature. But it's possible to implement this feature on top of the current functionality. I wouldn't address that in the current PR.

Reply:
Yep, I agree, it's for a separate PR. I'm curious what you folks think in general about reverts - is it needed or not. Because it's something that we haven't discussed when planning the feature.

Reply:
We'll definitely need a way to "revert" conflict resolutions. Just as an example: a rule update has an ABC scenario conflict for its `name` field. The user manually resolves the conflict, but then manually deletes the input value in the edit component. Since `name` is required, we should unresolve the conflict and disable the "Update rule" button.

But yes, we can add this complexity incrementally, doesn't need to be part of this PR.

Reply:
Continued the discussion in the Slack channel pasted above.